### PR TITLE
[12.0][MIG] sale_last_price_info

### PR DESCRIPTION
**Pull request by Swapnesh-SerpentCS**
**Product Last Price Info - Sale**
=======================

This module adds the last sale info of the product.

- Last Sale Price
- Last Sale Date
- Last Customer

### CODE REVIEW

File: sale_last_price_info/__init__.py
Comment (nikul-serpentcs) [inline sale_last_price_info/__init__.py]: @Swapnesh-SerpentCS Add new blank line

File: sale_last_price_info/tests/__init__.py
Comment (nikul-serpentcs) [inline sale_last_price_info/tests/__init__.py]: same here

**Review by nikul-serpentcs — COMMENTED** [review]
Minor Change

**Comment by Swapnesh-SerpentCS** [comment]
Commit Squashed and  Travis Done.
CC: @pedrobaeza

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM

File: sale_last_price_info/models/__init__.py
Comment (sudhir-erpharbor) [inline sale_last_price_info/models/__init__.py]: Please add a blank line at the EOF

**Review by sudhir-erpharbor — COMMENTED** [review]
Minor change

File: sale_last_price_info/models/product.py
Comment (sebalix) [inline sale_last_price_info/models/product.py]: It would be worth having a stored related field to get the date order at the line level, and use an order by directly with `search` + use the parameter `limit=1` to get the first. This field is going to be called quiet often, and the number of sale order lines for a given product can be huge as the time pass, PostgreSQL will handle that much better than Python.

File: sale_last_price_info/models/product.py
Comment (sebalix) [inline sale_last_price_info/models/product.py]: No need for the slice `:1` once the `search` call is improved.

**Review by sebalix — CHANGES_REQUESTED** [review]

File: sale_last_price_info/models/__init__.py
Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/__init__.py]: Done @sudhir-erpharbor

File: sale_last_price_info/models/product.py
Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/product.py]: That's what I thought but I decided to not change (Minimal diff policy) the behaviour (I know this doesn't change any functional behaviour)

However, I will add that field and Improve `_compute_last_sale` accordingly if we are getting same reviews from other @oca member too.

Comment (bizzappdev) [reply sale_last_price_info/models/product.py]: @Swapnesh-SerpentCS IMO best will be to search on sale.order with order='date_order desc' and limit=1 with product_id in the domain.

**Review by rven — CHANGES_REQUESTED** [review]
I'm missing the readme (https://github.com/OCA/maintainer-tools/tree/master/template/module/readme) section used by oca-gen-addon-readme
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: sale_last_price_info/models/product.py
Comment (rven) [reply sale_last_price_info/models/product.py]: I fully agree with the remark of @sebalix, in the beginning you won't notice the performance issue.
But when using the system for a couple of time, it will get slower and slower.
Maybe a new module should be added, a module that adds the date_order as related stored field on the sale order line. Then this can be used to solve this suggested change?

Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/product.py]: Hello @rven 
Hope you are doing well.

> Maybe a new module should be added, a module that adds the date_order as related stored field on the sale order line.

I think it's not good to add a new module just to add a new field (Without having additional usage) for one Purpose.
We can add a new module and add that in dependency if the field is going to be used in other modules too.

Comment (pedrobaeza) [reply sale_last_price_info/models/product.py]: No need for the stored field if you limit search to 1 and put the order on the `search` method directly (not after the results).

Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/product.py]: @pedrobaeza  Do you mean like this?
`search( [('product_id', '=', product.id), ('state', 'in', ['sale', 'done'])], limit=1).sorted( key=lambda l: l.order_id.date_order, reverse=True)`

Comment (rven) [reply sale_last_price_info/models/product.py]: @pedrobaeza this is not possible, because the field is only available on sale.order.
The search method does not work with order_id.date_order
`    def _check_qorder(self, word):
        if not regex_order.match(word):
            raise UserError(_('Invalid "order" specified. A valid "order" specification is a comma-separated list of valid field names (optionally followed by asc/desc for the direction)'))
        return True
`
To be able to use it in the search, it should be a database field defined on the sale.order.line itself.

Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/product.py]: I think we should go ahead with adding a related field in this module itself.

Comment (pedrobaeza) [reply sale_last_price_info/models/product.py]: You have 2 options then:

1. Add the date as related stored field in `sale.order.line`. This is convenient that the SQL column is created and populated in the pre-init-hook through an SQL for low installation time. And then do what I have proposed before. It's recommended also to put a new that can't conflict with other fields, like `date_order_sale_last_price_info` (I know, a bit cumbersome but it's the only way to assure uniqueness`, similar to what is done on onchanges).
2. Use the SQL view model `sale.report`, doing the search there, which is going to be more performant any way than current solution.

Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/product.py]: > @Swapnesh-SerpentCS IMO best will be to search on sale.order with order='date_order desc' and limit=1 with product_id in the domain.

@bizzappdev 
This would not be possible as we also need to get `price_unit` from related order line. (We have to use additional search for `sale.order.line where id=%s and product_id=%s` to get Order line if we use this way)

**Comment by Swapnesh-SerpentCS** [comment]
@bizzappdev @pedrobaeza @sebalix  @sudhir-erpharbor @rven 
This can be reviewed again.

File: sale_last_price_info/models/product.py
Comment (sebalix) [inline sale_last_price_info/models/product.py]: No need to use slice here, `line.` is enough.

File: sale_last_price_info/tests/test_sale_last_price_info.py
Comment (sebalix) [inline sale_last_price_info/tests/test_sale_last_price_info.py]: Same

**Review by sebalix — CHANGES_REQUESTED** [review]

File: sale_last_price_info/models/product.py
Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/models/product.py]: :+1:  and also this will work even if there is no line.

File: sale_last_price_info/tests/test_sale_last_price_info.py
Comment (Swapnesh-SerpentCS) [reply sale_last_price_info/tests/test_sale_last_price_info.py]: Done

**Review by bizzappdev — APPROVED** [review]
:+1:

**Review by sebalix — APPROVED** [review]

**Comment by Swapnesh-SerpentCS** [comment]
Thanks, @pedrobaeza.

However, I think still it will be Issue (Due to timezone) on converting datetime to date.
Eg, 

- Create a sale order with Datetime 22 Jan 2018 02:30 AM ((Indian Time  + 05:30))
- It will be stored as 21 Jan 2018 09:00 PM in UTC 
- So here it will be 21 Jan while 22 Jan is correct

Ps: I have not tested this in local yet.

**Pull request merged**
